### PR TITLE
mHC Learnable Residual Mixing: per-sublayer alpha/beta scale in TransolverBlock

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Transolver uses fixed residual connections `x_out = x + F(x)` in each block, with a fixed `skip_gate` that is initialized to approximate identity. This leaves no freedom for the optimizer to calibrate the **relative contribution of the skip path vs. the transformation path** on a per-block, per-sublayer basis.

Replacing fixed residuals with learnable scalar weights `x_out = alpha * x + beta * F(x)` (initialized at alpha=1, beta=1) gives 12 additional parameters (4 per block × 3 blocks) that the optimizer can use to:
1. Learn that early blocks (geometry encoding) need high skip weight (alpha near 1)
2. Learn that later blocks (global physics) may benefit from lower skip weight (alpha < 1, more transformation)
3. Enable per-block calibration for OOD generalization (NACA6416 vs NACA0012 may require different depth dynamics)

This is the **hyper-connection / learnable residual mixing** concept, explicitly requested by the human research team (GitHub Issue #1926). The safe (1, 1) initialization means epoch 0 is **identical to baseline** — zero regression risk.

**Expected improvement:** -1 to -3% p_oodc (OOD condition target), with possible p_tan improvement from better block depth calibration.

### 1. Add argparse flag

```python
parser.add_argument('--mhc_residuals', action='store_true', default=False,
                    help='Learnable alpha/beta residual weights per sublayer (hyper-connection mixing)')
```

### 2. Modify TransolverBlock

In `TransolverBlock.__init__`, add after the existing skip_gate:

```python
if getattr(cfg, 'mhc_residuals', False):
    # 4 learnable scalars per block: attn skip/transform + MLP skip/transform
    self.attn_alpha = nn.Parameter(torch.ones(1))   # attn sublayer: skip weight
    self.attn_beta  = nn.Parameter(torch.ones(1))   # attn sublayer: transform weight
    self.mlp_alpha  = nn.Parameter(torch.ones(1))   # MLP sublayer: skip weight
    self.mlp_beta   = nn.Parameter(torch.ones(1))   # MLP sublayer: transform weight
```

In `TransolverBlock.forward`, find the two residual additions (attention sublayer and MLP sublayer). Replace:

```python
# BEFORE (approximate identity via biased sigmoid skip_gate):
# fx = skip_gate(fx) * fx + attn_out

# AFTER:
if cfg.mhc_residuals:
    fx = self.attn_alpha * fx + self.attn_beta * attn_out
else:
    fx = skip_gate(fx) * fx + attn_out  # original

# Similarly for the MLP sublayer:
if cfg.mhc_residuals:
    fx = self.mlp_alpha * fx + self.mlp_beta * mlp_out
else:
    fx = skip_gate(fx) * fx + mlp_out  # original
```

**Implementation note:** Look at the actual `TransolverBlock.forward` code carefully — the exact variable names (`fx`, `attn_out`, `mlp_out`) may differ. The pattern to identify is the two `+` operations that add the sublayer output back to the running representation. Replace each with the alpha/beta version.

### 3. Monitor in W&B

Log the alpha/beta values for each block so we can see what the optimizer learns:
```python
# In training loop, log periodically:
for i, block in enumerate(model.blocks):
    if cfg.mhc_residuals:
        wandb.log({
            f'block{i}/attn_alpha': block.attn_alpha.item(),
            f'block{i}/attn_beta': block.attn_beta.item(),
            f'block{i}/mlp_alpha': block.mlp_alpha.item(),
            f'block{i}/mlp_beta': block.mlp_beta.item(),
        })
```

### 4. Run commands (2 seeds)

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent edward --seed 42 \
  --wandb_name "edward/mhc-residuals-s42" --wandb_group "edward/mhc-residuals" \
  --mhc_residuals \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature

# Seed 73: same with --seed 73 --wandb_name "edward/mhc-residuals-s73"
```

**Watch for:** If alpha or beta → 0 (one path killed), that is valid and interpretable. If val/loss diverges in epochs 1-3, try init alpha=1.0, beta=0.5 as a conservative alternative.

## Baseline

**Current best (PR #2213 — Wake Deficit Feature, 2-seed avg):**

| Metric | Value | Merge threshold |
|--------|-------|-----------------|
| p_in | **11.979** | < 11.98 |
| p_oodc | **7.643** | < 7.65 |
| p_tan | **28.341** | < 28.34 |
| p_re | **6.300** | < 6.30 |

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### Metrics (2-seed avg, seeds 42 & 73)

| Metric | Seed 42 | Seed 73 | **2-seed avg** | Baseline (PR #2213) | Delta |
|--------|---------|---------|---------------|---------------------|-------|
| p_in   | 13.151  | 12.193  | **12.672**    | 11.979              | +0.693 (+5.8%) ❌ |
| p_tan  | 29.225  | 28.164  | **28.695**    | 28.341              | +0.354 (+1.2%) ❌ |
| p_oodc | 7.796   | 7.691   | **7.744**     | 7.643               | +0.101 (+1.3%) ❌ |
| p_re   | 6.438   | 6.581   | **6.510**     | 6.300               | +0.210 (+3.3%) ❌ |

### Learned alpha/beta values (final epoch)

| Block | attn_α | attn_β | mlp_α | mlp_β |
|-------|--------|--------|-------|-------|
| 0 (s42/s73) | 1.96/1.93 | 0.11/0.13 | 1.76/1.76 | 0.26/0.27 |
| 1 (s42/s73) | 1.92/1.89 | 0.13/0.15 | 1.70/1.70 | 0.33/0.32 |
| 2 (s42/s73) | 1.86/1.88 | 0.18/0.15 | 1.76/1.77 | 0.27/0.26 |

### Run info

| | Seed 42 | Seed 73 |
|-|---------|---------|
| W&B run ID | k4xt5z1x | dh2taztr |
| Epochs completed | ~143 | ~145 |
| Runtime | 180.6 min (timeout) | 181.2 min (timeout) |

### Commands used

```bash
cd cfd_tandemfoil && python train.py --agent edward --seed 42 \
  --wandb_name "edward/mhc-residuals-s42" --wandb_group "edward/mhc-residuals" \
  --mhc_residuals \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
# Seed 73: same with --seed 73 --wandb_name "edward/mhc-residuals-s73"
```

### What happened

**NEGATIVE result.** All four surface pressure metrics regressed, with p_in worst at +5.8%.

The alpha/beta values reveal a clear and consistent pattern across both seeds and all blocks:
- **attn_alpha ≈ 1.9, attn_beta ≈ 0.13** → model nearly doubles the skip path and suppresses attention output to ~13%
- **mlp_alpha ≈ 1.7, mlp_beta ≈ 0.28** → similar but less extreme for MLP: skip boosted 70%, MLP output reduced to ~28%

The optimizer learned to **suppress the transformation paths** in favor of the skip connections. This is consistent across all 3 blocks and both seeds, meaning it's a stable learning dynamic, not noise. Crucially, the learned values are nearly identical across blocks — the hypothesis predicted per-block differentiation (early blocks favoring skip, later blocks favoring transformation), but in practice all blocks converge to the same skip-dominant solution.

**Why it failed:**

1. **The optimizer found a skip-dominant local minimum.** With unconstrained (α, β), gradient descent can easily move toward (2, 0) which is a near-identity mapping. The model effectively reduced its depth by downweighting all transformations, causing underfitting.

2. **Redundancy with existing mechanisms.** The baseline already has SE gating (`se = sigmoid(fc2(gelu(fc1(fx.mean(1)))))`) and DomainLayerNorm that provide adaptive feature calibration. The mHC scalars compete with these mechanisms rather than complementing them.

3. **No per-head granularity.** A single scalar per sublayer is too coarse — the model needs to boost some attention heads while suppressing others, not uniformly scale the entire sublayer output.

### Suggested follow-ups

1. **Constrain alpha ∈ [0.5, 1.5] and beta ∈ [0.5, 1.5]** with sigmoid reparameterization to prevent skip-dominant collapse.
2. **Per-head mixing** (alpha/beta per attention head rather than per sublayer) for finer-grained control.
3. **Remove SE gating** when using mHC — the two mechanisms appear to interfere (SE already provides an adaptive gate).
4. **Initialize alpha=0.5, beta=1.0** (favor transformation over skip) to counteract the skip-dominant learning dynamic.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2222 because All 4 metrics regressed (p_in +5.8%, p_tan +1.2%, p_oodc +1.3%, p_re +3.3%). The learned alpha/beta values are very informative: the optimizer consistently converged to alpha≈1.8-1.9, beta≈0.1-0.3 across ALL blocks and BOTH seeds — meaning it learned to suppress all transformation paths and boost skip connections. The model effectively reduced its own depth, causing underfitting. This is a stable, consistent finding that tells us the Transolver backbone already has well-calibrated residual dynamics via its existing SE gating and DomainLayerNorm. Learnable residual mixing without constraints collapses to skip-dominant solutions. Excellent analysis of the learned values — the per-block alpha/beta data is genuinely useful for understanding the architecture. Direction exhausted (unconstrained mHC). The human team's intuition (issue #1926) was reasonable but the existing SE gating already provides the adaptive calibration that mHC targets..
